### PR TITLE
Partial points scoring

### PR DESCRIPTION
**Pull request by MattyMay**
Add partial credit scoring capabilities

### CODE REVIEW

File: autograder/core/submission_feedback.py [hunk 1]
```diff
@@ -1190,20 +1210,64 @@ def stderr_points_possible(self) -> int:
...
+    def partial_credit_points(self) -> int:
+        if not self._fdbk.show_points:
+            return 0
+        if self._cmd.partial_credit_source == PartialCreditSource.none:
+            return 0
+        if (self._ag_test_command_result.partial_credit_points
+                > self._cmd.max_points_for_partial_credit):
+            return self._cmd.max_points_for_partial_credit
+
+        return self._ag_test_command_result.partial_credit_points
```
Comment (james-perretta) [inline autograder/core/submission_feedback.py]: Could this use `min` instead?

File: autograder/core/submission_feedback.py [hunk 2]
```diff
@@ -1190,20 +1210,64 @@ def stderr_points_possible(self) -> int:
...
+        return self._cmd.max_points_for_partial_credit
+
+    @property
+    def partial_credit_error(self) -> str:
+        if not self._fdbk.show_points:
+            return 'none'
+
+        match self._ag_test_command_result.partial_credit_error:
+            case PartialCreditError.none:
+                return 'none'
```
Comment (james-perretta) [inline autograder/core/submission_feedback.py]: Maybe this should just be an empty string? I could see this getting interpreted as "there's an error that has to do with None"

File: autograder/core/tests/test_models/test_ag_test/test_ag_test_command.py [hunk 3]
```diff
@@ -798,3 +807,36 @@ def test_stdin_stdout_stderr_source_only_set_to_none_if_source_is_instructor_fil
         self.assertEqual(ag_models.ExpectedOutputSource.text,
                          self.ag_test_command.expected_stderr_source)
         self.assertIsNone(self.ag_test_command.expected_stderr_instructor_file)
+
+    def test_partial_credit_source_stdout_when_expected_stdout_source_is_not_none(self):
```
Comment (james-perretta) [inline autograder/core/tests/test_models/test_ag_test/test_ag_test_command.py:811]: I usually include a check that the appropriate field is in the validation error dictionary. Search the codebase for `as cm:` to see an example

File: autograder/core/tests/test_models/test_ag_test/test_ag_test_command.py [hunk 4]
```diff
@@ -798,3 +807,36 @@ def test_stdin_stdout_stderr_source_only_set_to_none_if_source_is_instructor_fil
...
+            )
+
+    def test_partial_credit_source_stderr_when_expected_stderr_source_is_not_none(self):
+        with self.assertRaises(exceptions.ValidationError):
+            self.ag_test_command.validate_and_update(
+                expected_stderr_source=ag_models.ExpectedOutputSource.text,
+                partial_credit_source=ag_models.PartialCreditSource.stderr
+            )
+
+    def test_partial_credit_source_stdout_when_expected_stderr_source_is_not_none(self):
```
Comment (james-perretta) [inline autograder/core/tests/test_models/test_ag_test/test_ag_test_command.py:827]: Can we add a test for the fourth combination? (partial credit source stderr, expected stdout not none)

File: autograder/core/tests/test_submission_feedback/test_ag_test_command_result_feedback.py [hunk 5]
```diff
@@ -140,6 +191,136 @@ def test_points_everything_incorrect_max_fdbk(self):
...
+                         fdbk.partial_credit_points_possible)
+
+        self.assertEqual(self.min_points_possible + 1, fdbk.total_points)
+        self.assertEqual(self.max_points_possible, fdbk.total_points_possible)
+
+    def test_points_negative_stdout_partial_credit_points(self):
+        self.ag_test_command = obj_build.make_stdout_partial_credit_test_command()
+        cmd_result = self.make_incorrect_result()
+
+        cmd_result.partial_credit_points = - 42
```
Comment (james-perretta) [inline autograder/core/tests/test_submission_feedback/test_ag_test_command_result_feedback.py]: Formatting: remove space between - and 42

File: autograder/grading_tasks/tasks/grade_ag_test.py [hunk 6]
```diff
@@ -258,6 +259,34 @@ def grade_ag_test_command_impl(sandbox: AutograderSandbox,
...
             result_data['stderr_correct'] = diff.diff_pass
 
+        if (ag_test_cmd.partial_credit_source != ag_models.PartialCreditSource.none
+                and run_result.timed_out is False):
+            if ag_test_cmd.partial_credit_source == ag_models.PartialCreditSource.stdout:
+                partial_credit_source = run_result.stdout
+            else:
+                partial_credit_source = run_result.stderr
+
+            with open(partial_credit_source.name) as f:
```
Comment (james-perretta) [inline autograder/grading_tasks/tasks/grade_ag_test.py]: Note to self: discuss caveats

File: autograder/core/tests/test_submission_feedback/test_ag_test_command_result_feedback.py [hunk 7]
```diff
@@ -140,6 +195,148 @@ def test_points_everything_incorrect_max_fdbk(self):
...
+
+        self.assertEqual(self.min_points_possible, fdbk.total_points)
+        self.assertEqual(self.max_points_possible, fdbk.total_points_possible)
+
+    def test_points_everything_incorrect_with_some_stdout_partial_credit_points(self):
+        self.ag_test_command = obj_build.make_stdout_partial_credit_test_command(
+            self.ag_test_case,
+            max_points_for_partial_credit=2
+        )
+        cmd_result = self.make_incorrect_result()
```
Comment (james-perretta) [inline autograder/core/tests/test_submission_feedback/test_ag_test_command_result_feedback.py:241]: Optional: make this function (and the underlying make_incorrect_ag_test_command_result that it calls) take in an argument to specify partial credit points (default to 0)

File: autograder/core/tests/test_submission_feedback/test_ag_test_command_result_feedback.py [hunk 8]
```diff
@@ -949,6 +1146,54 @@ def test_student_on_fail_description_shown_on_fail(self) -> None:
             stderr_correct=False,
         )
 
+    def test_student_on_fail_description_shown_on_fail_with_partial_credit(self):
+        description = 'description WAAAAAA'
+        on_fail_description = 'WAAAA fail'
+        normal_fdbk_config = self.ag_test_command.normal_fdbk_config
```
Comment (james-perretta) [inline autograder/core/tests/test_submission_feedback/test_ag_test_command_result_feedback.py:1188]: I think this could be passed to the call to `make_stdout_partial_credit_test_command` as a kwarg instead

File: autograder/core/tests/test_submission_feedback/test_ag_test_command_result_feedback.py [hunk 9]
```diff
@@ -949,6 +1146,54 @@ def test_student_on_fail_description_shown_on_fail(self) -> None:
...
+        opts = (True, False, None)
+
+        self.ag_test_command = obj_build.make_stdout_partial_credit_test_command(
+            self.ag_test_case,
+            max_points_for_partial_credit=1
+        )
+        self.ag_test_command.validate_and_update(normal_fdbk_config=normal_fdbk_config)
+        for (
+            return_code_correct, stderr_correct, partial_credit_correct
+        ) in itertools.product(opts, opts, opts):
```
Comment (james-perretta) [inline autograder/core/tests/test_submission_feedback/test_ag_test_command_result_feedback.py]: I think it's fine to inline the argument to `itertools.product` in both places--makes it clearer what their purpose is. `itertools.product(*itertools.repeat([True, False, None], 3))` should do the trick, but I'm open to alternatives

File: autograder/grading_tasks/tests/test_tasks/test_grade_ag_test.py [hunk 10]
```diff
@@ -237,6 +237,94 @@ def test_wrong_expected_stderr_instructor_file(self, *args):
...
+
+    def test_partial_credit_more_than_max_points(self, *args):
+        cmd = obj_build.make_stdout_partial_credit_test_command(
+            self.ag_test_case,
+            max_points_for_partial_credit=1,
+            cmd='printf "<!! score: 2 !!>"')
+        tasks.grade_submission_task(self.submission.pk)
+
+        res = ag_models.AGTestCommandResult.objects.get(ag_test_command=cmd)
+        self.assertEqual(res.partial_credit_points, 2)
```
Comment (james-perretta) [inline autograder/grading_tasks/tests/test_tasks/test_grade_ag_test.py:322]: Reminder: add tests to ag test command feedback for taking the max if the score reported goes over

File: autograder/grading_tasks/tests/test_tasks/test_grade_ag_test.py [hunk 11]
```diff
@@ -400,6 +488,28 @@ def test_program_times_out(self, *args):
         res = ag_models.AGTestCommandResult.objects.get(ag_test_command=cmd)
         self.assertTrue(res.timed_out)
 
+    def test_program_times_out_with_stdout_partial_credit(self, *args):
+        cmd = obj_build.make_stdout_partial_credit_test_command(
+            self.ag_test_case,
+            cmd=self.timeout_cmd,
+            time_limit=1)
```
Comment (james-perretta) [inline autograder/grading_tasks/tests/test_tasks/test_grade_ag_test.py:534]: Set max partial credit points to nonzero

File: autograder/grading_tasks/tests/test_tasks/test_grade_ag_test.py [hunk 12]
```diff
@@ -400,6 +488,28 @@ def test_program_times_out(self, *args):
...
+        tasks.grade_submission_task(self.submission.pk)
+
+        res = ag_models.AGTestCommandResult.objects.get(ag_test_command=cmd)
+        self.assertTrue(res.timed_out)
+        self.assertEqual(res.partial_credit_error, ag_models.PartialCreditError.none)
```
Comment (james-perretta) [inline autograder/grading_tasks/tests/test_tasks/test_grade_ag_test.py:539]: Add check that partial credit points are zero

File: autograder/grading_tasks/tests/test_tasks/test_grade_ag_test.py [hunk 13]
```diff
@@ -400,6 +488,28 @@ def test_program_times_out(self, *args):
...
+
+    def test_program_times_out_with_stderr_partial_credit(self, *args):
+        cmd = obj_build.make_stderr_partial_credit_test_command(
+            self.ag_test_case,
+            cmd=self.timeout_cmd,
+            time_limit=1)
```
Comment (james-perretta) [inline autograder/grading_tasks/tests/test_tasks/test_grade_ag_test.py:546]: Same as above

File: autograder/utils/testing/model_obj_builders.py [hunk 14]
```diff
@@ -412,6 +412,63 @@ def make_full_ag_test_command(
     return ag_models.AGTestCommand.objects.validate_and_create(**base_kwargs)
 
 
+def make_stdout_partial_credit_test_command(
+    ag_test_case: ag_models.AGTestCase | None = None,
+    max_points_for_partial_credit: int | None = None,
```
Comment (james-perretta) [inline autograder/utils/testing/model_obj_builders.py]: Suggestion: have this default to some non-zero number, since everywhere this gets called passes in a value

File: autograder/core/tests/test_submission_feedback/test_ag_test_command_result_feedback.py [hunk 15]
```diff
@@ -140,6 +195,148 @@ def test_points_everything_incorrect_max_fdbk(self):
...
+        fdbk = get_cmd_fdbk(cmd_result, ag_models.FeedbackCategory.max)
+
+        self.assertEqual(-1337,
+                         fdbk.partial_credit_points)
+        self.assertEqual(self.ag_test_command.max_points_for_partial_credit,
+                         fdbk.partial_credit_points_possible)
+
+        self.assertEqual(self.min_points_possible - 1337, fdbk.total_points)
+        self.assertEqual(self.max_points_possible, fdbk.total_points_possible)
+
```
Comment (james-perretta) [inline autograder/core/tests/test_submission_feedback/test_ag_test_command_result_feedback.py:325]: Add test for points hidden? (check points, max points, and error)

File: autograder/core/submission_feedback.py [hunk 16]
```diff
@@ -1190,20 +1210,64 @@ def stderr_points_possible(self) -> int:
...
+            case PartialCreditError.non_integer:
+                return 'Non-integer value found for partial credit score'
+            case PartialCreditError.exceeded_max_points:
+                return ('Partial credit points exceeded the number of partial credit'
```
Comment (james-perretta) [inline autograder/core/submission_feedback.py]: Idea: include the points parsed in this message?

**Review by james-perretta — APPROVED** [review]
